system/lib/sbrk.c

```diff
@@ -29,6 +29,23 @@
 #define SET_ERRNO()
 #endif
 
+static intptr_t sbrk_val = 0;
```

Comment:
Can you just do `static intptr_t  sbrk_val = (intptr_t)&__heap_base` right here at least in the non-pthread case?

In fact, even in the pthread case I should work since it will generate an assignment from `GOT.mem.__heap_base` on startup

Reply:
That can work, but is less good for code size it seems. The issue is that `static intptr_t sbrk_val = (intptr_t)&__heap_base;` adds that number into the memory segment initializer. If wasm-opt then manages to remove malloc and sbrk, it still can't remove that. So super-tiny programs end up almost empty, but with a couple of bytes in a memory segment for no obvious reason. We have a test that fails on this, in fact, the metadce test that exports nothing and expects to get an empty wasm file with no data and no functions.

So it's better for super-tiny programs to do it this way. And for large programs that use malloc, it doesn't matter much either way we do it. But, I'm not totally happy with this I admit...

Reply:
I have quite a strong preference for `static intptr_t sbrk_val = (intptr_t)&__heap_base;` ... since it lets the linker do the work of assign the correct value.

Also, in smaller programs that don't reference `sbrk_val` at all the linker can eliminate this global completely, no?

This kind of micro optimization doesn't seem worth it.  There is also the cost in terms of wasm code size because the size of you emscripten_get_sbrk_ptr is bigger that just:

```
return &sbrk_val
```
```

Reply:
> Also, in smaller programs that don't reference sbrk_val at all the linker can eliminate this global completely, no?

No, because we link in malloc by default, and only remove it through metadce when possible.

I think it would be good to move towards a model that does not link in malloc by default, and can then do it the way we both prefer, but that's not trivial and I think should be kept separate from this. For now, I'd like to not regress code size or remove the existing test that breaks unless this is properly optimized.

(It's true this has a small cost in code size when malloc is included. But if malloc is there then it's a tiny addition to the much larger malloc; and if malloc isn't there, the saved code size with this optimization is much more noticeable.)

Reply:
I'm still not convinced...  the only programs that can benefit here are ones that don't use the heap at all (not just malloc... but any heap usage at all right?)

Do you think we need to be optimizing for such programs?

How about we add an option to not export malloc by default that such (very unusual) programs could use to get this saving?

Reply:
MINIMAL_RUNTIME and STANDALONE_WASM programs already don't export malloc by default so won't be effected.

Reply:
I guess I'm OK with the TODO here... but I'd perhaps rather invert it by added a TODO to emcc.py line 1408 that says "TODO: exporting malloc here causes sbrk_ptr to be included and it cannot be removed by metadce. Find a way to avoid exporting malloc by default?" 

:)

But lgtm this way too if you feel strongly.

Reply:
> I'm still not convinced... the only programs that can benefit here are ones that don't use the heap at all (not just malloc... but any heap usage at all right?) Do you think we need to be optimizing for such programs?

Yes, I think it's nice that we emit a tiny wasm for
```wat
(module
  (func "double_it" (param $x i32)
    (i32.mul (i32.const 2) (local.get $x))))
```
Such pure computational stuff shouldn't have data segments, and it would be a regression to change that.

But the good news is that I have a PR almost ready for this, in the `nomalloc` branch! So I don't think any TODO comment is needed here. It passes tests and shrinks code size, but I'll do some more careful work to try to avoid regressions before I open it.

Reply:
sgtm